### ISSUE TITLE
Add support for "Cache, Update and Refresh" caching strategy

### ISSUE THREAD

**Issue by ChrisBAshton** [ampproject/amphtml#21649]
## Describe the new feature or change to an existing feature you'd like to see

Whilst AMP has some support for Service Workers via the [amp-install-serviceworker](https://ampbyexample.com/components/amp-install-serviceworker/) component, interaction between the service worker and the window is extremely limited.

I currently have a [Cache and Update](https://serviceworke.rs/strategy-cache-and-update.html) strategy, which means my AMP application serves from cache by default and updates in the background. Ideally, I'd like to switch to [Cache, Update and Refresh](https://serviceworke.rs/strategy-cache-update-and-refresh.html) so that if the page and the cache are out of sync, I can tell the user and give them the option to refresh (e.g. "The content of this page has updated. [Click to refresh](#)).

The 'native' way of doing this is to send `postMessage` events to the window, like so:

```js
return self.clients.matchAll().then(function (clients) {
  clients.forEach(function (client) {
    var message = {
      type: 'refresh',
      url: response.url
    };
    client.postMessage(JSON.stringify(message));
  });
  return response;
});
```

...which has a corresponding listener on the front-end:

```js
function showRefreshMsgToUser() {
    if ('serviceWorker' in navigator) {
        navigator.serviceWorker.onmessage = function (event) {
            var message = JSON.parse(event.data);
            if (message.type === 'refresh') {
                var msg = 'This page has been updated, <a href="'+message.url+'">click here to refresh it</a>.';
                var alert = '<div class="alert alert-warning alert-dismissible" role="alert"><button type="button" class="close" aria-hidden="true" data-dismiss="alert">×</button>'+msg+'</div>';
                jQuery('body').prepend(alert);
            }
        };
    }
}
```

Example from [Implementing A Service Worker For Single-Page App WordPress Sites](https://www.smashingmagazine.com/2017/10/service-worker-single-page-application-wordpress-sites/).

But as AMP doesn't support inline JavaScript, there is no way of hooking into the postMessage event (apart from perhaps hacking around the issue with [amp-iframe](https://www.ampproject.org/docs/reference/components/amp-iframe)).

I'd like a way of implementing `Cache, Update and Refresh` - perhaps as a configurable event on the `amp-install-serviceworker` component?

## Describe alternatives you've considered

N/A

## Additional context

Possibly related issues: #4479, #9074.

**Comment by torch2424** [comment]
Triaging to @prateekbh , Feel free to re-assign 😄

cc @jridgewell while @prateekbh is out. 👍

**Comment by jpettitt** [comment]
Maybe enhance `amp-user-notification` to listen for message from a service worker?

**Comment by prateekbh** [comment]
An effort to standardize the effort of communication between workers and AMP-state is being tracked in #18615. That should in this use case as well

**Comment by prateekbh** [comment]
probably can be closed as duplicate of #18615

**Comment by ChrisBAshton** [comment]
Happy to close in favour of #18615.
